athena-gcs/src/main/java/com/amazonaws/athena/connectors/gcs/storage/StorageMetadata.java

```diff
@@ -0,0 +1,286 @@
+        String path = locationUri.getPath().startsWith("/") ? locationUri.getPath().substring(1) : locationUri.getPath();
+        Page<Blob> blobs = storage.list(bucketName, prefix(path));
+        for (Blob blob : blobs.iterateAll()) {
+            if (isBlobFile(blob)) {
+                fileList.add(bucketName + "/" + blob.getName());
+            }
+        }
+        return fileList;
+    }
+
+    /**
+     * Retrieves a list of partition folders from the GCS bucket based on partition.pattern Table parameter and partition keys set forth in Glue table. If the summary from the
+     * constraints is empty (no where clauses or unsupported clauses), it will essentially return all the partition folders from the GCS bucket. If there is any constraints to
+     * apply, it will apply constraints to filter selected partition folder, to narrow down the data load
+     *
+     * @param schema An instance of {@link Schema} that describes underlying Table's schema
+     * @param tableInfo Name of the table
+     * @param constraints An instance of {@link Constraints}, captured from where clauses
+     * @param awsGlue An instance of {@link AWSGlue}
+     * @return A list of {@link List<Map<String, String>>} instances
+     * @throws URISyntaxException Throws if any occurs during parsing Uri
+     */
+    public Set<List<Map<String, String>>> getPartitionFolders(Schema schema, TableName tableInfo, Constraints constraints, AWSGlue awsGlue)
+            throws URISyntaxException
+    {
+        LOGGER.info("Getting partition folder(s) for table {}.{}", tableInfo.getSchemaName(), tableInfo.getTableName());
+        Set<List<Map<String, String>>> partitionFolders = new HashSet<>();
+        Table table = GcsUtil.getGlueTable(tableInfo, awsGlue);
+
+        List<AbstractExpression> expressions = new FilterExpressionBuilder(schema).getExpressions(constraints);
+        LOGGER.info("Expressions for the request of {}.{} is \n{}", tableInfo.getSchemaName(), tableInfo.getTableName(), expressions);
+        URI storageLocation = new URI(table.getStorageDescriptor().getLocation());
+        LOGGER.info("Listing object in location {} under the bucket {}", storageLocation.getAuthority(), storageLocation.getPath());
+        String path = storageLocation.getPath().substring(1);
+        Page<Blob> blobPage = storage.list(storageLocation.getAuthority(), prefix(path));
+        for (Blob blob : blobPage.iterateAll()) {
+            String folderPath = blob.getName().replaceFirst("^" + path, "");
+            // remove the front-slash, because, the expression generated without it
+            if (folderPath.startsWith("/")) {
+                folderPath = folderPath.substring(1);
+            }
+            LOGGER.info("Examining folder {}", folderPath);
+            List<Map<String, String>> partitions = PartitionUtil.getPartitionColumnData(table, folderPath);
+            if (!partitions.isEmpty() && checkPartitionWithConstrains(partitions, expressions)) {
```

Comment:
I already mentioned this in my previous comment, but I'll mention it again:

`checkPartitionWithConstrains` is being called in a loop, so this means that the `expressionMap` that its building inside that method call is going to be rebuilt over and over again.

Instead, you can just build the `expressionMap` after line 150 and then just pass it in, like:
`checkPartitionWithConstrains(partitions, expressionMap)`.

To be explicit, just move lines 217-219 to be after line 150, and then just pass in the `expressionMap` into `checkPartitionWithConstrains()`.

Reply:
I just fixed this for you here: https://github.com/trianz-azam/aws-athena-query-federation/commit/ab6d61fbecbff187ae7aec454c4eb719c7da5009